esp-hal/src/gpio/mod.rs

```diff
@@ -961,80 +967,64 @@ where
     }
 }
 
-/// General Purpose Input/Output driver
-pub struct Io {
-    _io_mux: IO_MUX,
-    /// The pins available on this chip
-    pub pins: Pins,
-}
-
-impl Io {
-    /// Initialize the I/O driver.
-    pub fn new(gpio: GPIO, io_mux: IO_MUX) -> Self {
-        Self::new_with_priority(gpio, io_mux, crate::interrupt::Priority::min())
-    }
-
-    /// Initialize the I/O driver with a interrupt priority.
-    ///
-    /// This decides the priority for the interrupt when using async.
-    pub fn new_with_priority(
-        mut gpio: GPIO,
-        io_mux: IO_MUX,
-        prio: crate::interrupt::Priority,
-    ) -> Self {
-        gpio.bind_gpio_interrupt(gpio_interrupt_handler);
-        crate::interrupt::enable(crate::peripherals::Interrupt::GPIO, prio).unwrap();
-
-        Self::new_no_bind_interrupt(gpio, io_mux)
-    }
-
-    /// Initialize the I/O driver without enabling the GPIO interrupt or
-    /// binding an interrupt handler to it.
-    ///
-    /// *Note:* You probably don't want to use this, it is intended to be used
-    /// in very specific use cases. Async GPIO functionality will not work
-    /// when instantiating `Io` using this constructor.
-    pub fn new_no_bind_interrupt(gpio: GPIO, _io_mux: IO_MUX) -> Self {
-        Io {
-            _io_mux,
-            pins: gpio.pins(),
+impl<const GPIONUM: u8> InterruptConfigurable for GpioPin<GPIONUM>
+where
+    Self: GpioProperties,
+{
+    fn set_interrupt_handler(&mut self, handler: InterruptHandler) {
+        raise_gpio_interrupt_priority(handler.priority());
+        USER_INTERRUPT_HANDLERS[GPIONUM as usize].store(handler.handler());
+    }
+}
+
+static MAX_PRIO: AtomicUsize = AtomicUsize::new(Priority::None as usize);
+
+fn raise_gpio_interrupt_priority(priority: Priority) {
+    #[cold]
+    fn do_raise_priority_level(current_prio: usize, priority: Priority) {
```

Comment:
I don't think you need atomics for this since handler should only be bound to one core. If you think otherwise then we have bigger fish to fry now.

Reply:
I guess we're frying the bigger fish then.
So how are multi core interrupts supposed to work here, which core handles which pin, etc.
Is the interrupt handler core safe?

I've always thought of all the async drivers as `!Send` since you don't have to worry about core/thread safety which is not as cheap to deal with compared to sticking to a single core.